src/decoder/mod.rs

```diff
@@ -188,217 +576,255 @@ impl<R: Read + Seek> DecoderImpl<R> {
             #[cfg(all(feature = "minimp3", not(feature = "symphonia-mp3")))]
             DecoderImpl::Mp3(source) => source.try_seek(pos),
             #[cfg(feature = "symphonia")]
-            DecoderImpl::Symphonia(source) => source.try_seek(pos),
-            DecoderImpl::None(_) => Err(SeekError::NotSupported {
-                underlying_source: "DecoderImpl::None",
-            }),
+            DecoderImpl::Symphonia(source, PhantomData) => source.try_seek(pos),
         }
     }
 }
 
-impl<R> Decoder<R>
-where
-    R: Read + Seek + Send + Sync + 'static,
-{
-    /// Builds a new decoder.
-    ///
-    /// Attempts to automatically detect the format of the source of data.
-    #[allow(unused_variables)]
-    pub fn new(data: R) -> Result<Decoder<R>, DecoderError> {
-        #[cfg(all(feature = "wav", not(feature = "symphonia-wav")))]
-        let data = match wav::WavDecoder::new(data) {
-            Err(data) => data,
-            Ok(decoder) => {
-                return Ok(Decoder(DecoderImpl::Wav(decoder)));
-            }
-        };
-
-        #[cfg(all(feature = "flac", not(feature = "symphonia-flac")))]
-        let data = match flac::FlacDecoder::new(data) {
-            Err(data) => data,
-            Ok(decoder) => {
-                return Ok(Decoder(DecoderImpl::Flac(decoder)));
-            }
-        };
-
-        #[cfg(all(feature = "vorbis", not(feature = "symphonia-vorbis")))]
-        let data = match vorbis::VorbisDecoder::new(data) {
-            Err(data) => data,
-            Ok(decoder) => {
-                return Ok(Decoder(DecoderImpl::Vorbis(decoder)));
-            }
-        };
-
-        #[cfg(all(feature = "minimp3", not(feature = "symphonia-mp3")))]
-        let data = match mp3::Mp3Decoder::new(data) {
-            Err(data) => data,
-            Ok(decoder) => {
-                return Ok(Decoder(DecoderImpl::Mp3(decoder)));
-            }
-        };
+/// Converts a `File` into a `Decoder` with automatic optimizations.
+///
+/// This implementation:
+/// - Wraps the file in a `BufReader` for better performance
+/// - Sets `byte_len` from file metadata when available, which can improve seeking operations
+///
+/// # Examples
+/// ```no_run
+/// use std::fs::File;
+/// use std::convert::TryFrom;
+/// use rodio::Decoder;
+///
+/// let file = File::open("audio.mp3").unwrap();
+/// let decoder = Decoder::try_from(file).unwrap();
+/// ```
+impl TryFrom<std::fs::File> for Decoder<BufReader<std::fs::File>> {
+    type Error = DecoderError;
 
-        #[cfg(feature = "symphonia")]
-        {
-            let mss = MediaSourceStream::new(
-                Box::new(ReadSeekSource::new(data)) as Box<dyn MediaSource>,
-                Default::default(),
-            );
+    fn try_from(file: std::fs::File) -> Result<Self, Self::Error> {
+        let mut builder = DecoderBuilder::new();
+        if let Some(len) = file.metadata().ok().map(|m| m.len()) {
```

Comment:
We could return an error if the metadata is not available. If we can play the file we should be able to get the metadata for it. In (I would say extreme) edge cases where there is no metadata and for some reason the file is playable users can now use the builder :tada: .

Reply:
I agree. Done.